tools/clang/lib/SPIRV/EmitSPIRVAction.cpp

```diff
@@ -422,6 +444,56 @@ class SPIRVEmitter : public ASTConsumer {
     }
   }
 
+  spv::Op translateOp(BinaryOperator::Opcode op, QualType type) {
+    // TODO: the following is not considering vector types yet.
+    const bool isSintType = type->isSignedIntegerType();
+    const bool isUintType = type->isUnsignedIntegerType();
+    const bool isFloatType = type->isFloatingType();
+
+#define BIN_OP_CASE_INT_FLOAT(kind, intBinOp, floatBinOp)                      \
+  \
+case BO_##kind : {                                                             \
+    if (isSintType || isUintType) {                                            \
+      return spv::Op::Op##intBinOp;                                            \
+    }                                                                          \
+    if (isFloatType) {                                                         \
+      return spv::Op::Op##floatBinOp;                                          \
+    }                                                                          \
+  }                                                                            \
+  break
+
+#define BIN_OP_CASE_SINT_UINT_FLOAT(kind, sintBinOp, uintBinOp, floatBinOp)    \
+  \
+case BO_##kind : {                                                             \
+    if (isSintType) {                                                          \
+      return spv::Op::Op##sintBinOp;                                           \
+    }                                                                          \
+    if (isUintType) {                                                          \
+      return spv::Op::Op##uintBinOp;                                           \
+    }                                                                          \
+    if (isFloatType) {                                                         \
+      return spv::Op::Op##floatBinOp;                                          \
+    }                                                                          \
+  }                                                                            \
+  break
+
+    switch (op) {
+      BIN_OP_CASE_INT_FLOAT(Add, IAdd, FAdd);
+      BIN_OP_CASE_INT_FLOAT(Sub, ISub, FSub);
+      BIN_OP_CASE_INT_FLOAT(Mul, IMul, FMul);
+      BIN_OP_CASE_SINT_UINT_FLOAT(Div, SDiv, UDiv, FDiv);
+      BIN_OP_CASE_SINT_UINT_FLOAT(Rem, SMod, UMod, FMod);
```

Comment:
You are right. I don't have a clear answer for this issue now.

SPIR-V defines both `Op*Rem` and `Op*Mod` for "remainder" operations but without specify exactly how they should be calculated. I think it assumes some general understanding of the "remainder" operation, for which my Googlefu failed and I cannot find a definite guide. What I have right now is [`OpFRem` tests](https://github.com/KhronosGroup/VK-GL-CTS/blob/master/external/vulkancts/modules/vulkan/spirv_assembly/vktSpvAsmInstructionTests.cpp#L803) in dEQP uses [`std::fmod()`](http://en.cppreference.com/w/cpp/numeric/math/fmod), and the dEQP tests for `OpSRem` and `OpSMod` both saying [negative operands result in undefined behavior](https://github.com/KhronosGroup/VK-GL-CTS/blob/master/external/vulkancts/modules/vulkan/spirv_assembly/vktSpvAsmInstructionTests.cpp#L1074). There are also a [wiki page](https://en.wikipedia.org/wiki/Remainder) and a [MathForum thread](http://mathforum.org/library/drmath/view/52343.html) about this. But I'm not sure SPIR-V actually means the same thing.

Right now using `OpSMod` to be consistent with glslang; but clang DXC and C's handling of `%` suggest I should go with `Op*Rem`.